include/exception.h

```diff
@@ -105,32 +107,36 @@ typedef struct
     volatile reg_block_t* regs;
 } exception_t;
 
+/** @} */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
 
 /** 
- * @brief Guaranteed length of the reset time.
+ * @brief Generic exception handler
  * 
- * This is the guaranteed length of the reset time, that is the time
- * that goes between the user pressing the reset button, and the CPU actually
- * resetting. See #exception_reset_time for more details.
+ * This is the type of a handler that can be registered using #register_exception_handler.
+ * It is associated to all unhandled exceptions that are not otherwise handled by libdragon.
  * 
- * @note The general knowledge about this is that the reset time should be
- *       500 ms. Testing on different consoles show that, while most seem to
- *       reset after 500 ms, a few EU models reset after 200ms. So we define
- *       the timer shorter for greater compatibility.
+ * @param exc Exception information
  */
-#define RESET_TIME_LENGTH      TICKS_FROM_MS(200)
-
-/** @} */
+typedef void (*exception_handler_t)(exception_t *exc);
 
-#ifdef __cplusplus
-extern "C" {
-#endif
+/** 
+ * @brief Syscall handler
+ * 
+ * This is the type of a handler of a syscall exception.
+ * 
+ * @param exc Exception information
+ * @param code Syscall code
+ */
+typedef void (*syscall_handler_t)(exception_t *exc, uint32_t code);
 
-void register_exception_handler( void (*cb)(exception_t *) );
+exception_handler_t register_exception_handler( exception_handler_t cb );
 void exception_default_handler( exception_t* ex );
 
-void register_reset_handler( void (*cb)(void) );
-uint32_t exception_reset_time( void );
+void register_syscall_handler( syscall_handler_t cb, uint32_t first_code, uint32_t last_code );
```

Comment:
I actually had implemented with a mask before, but it was harder to read and use. The idea is that if you are going to use a syscall, you might need a few variations around it. For instance, the inspector in the unstable branch currently uses 3 syscalls which are three different entry points (MIPS exceptions, assertions, C++ exceptions). A range is simpler to understand and read in my opinion.

Reply:
Might be easier to read but I thought we can save a few ops. From readability standpoint, ranges might be confusing as well b/c of inclusive/exclusive ranges and not possible to represent ranges with holes in it. I don't have a too strong of an opinion though.

Reply:
Speedwise, a few cycles per syscall isn't a problem, we don't have any plan to use syscalls in code sensitive codepaths at the moment. Consider that a range is checked with:

```
  if (code >= min && code <= max)
```

which gets compiled to

```
  if (code-min >= max-min)
```
so it's 3 cycles.

A mask is:

```
  if (code & mask == value)
```

which is 2 cycles. Honestly not a big deal.

I have two other alternatives in mind:
1) Not declare a range, and invoke all syscall handlers always with the code. Each handler will check if they like the code and do something. This is the simplest API-wise but we don't get to detect conflicts.
2) Do like we do in rspq and settle on a fixed split. Since the syscall code is 20 bit, we could decide to split 8+12 for instance, and simply let register on a single 8 bit prefix, and forward to it all codes with that prefix.

Reply:
Honestly this API isn't going to be used so much I think. It could even be kept private for now, I doubt there are many use cases for libraries to use syscalls. 

Basically the main advantage of a syscall handler is that you get a snapshot of the state of all registers at that point, so it must be some advanced / debugging feature. We have two uses for now: one is the inspector (which, being a "debugger", requires a snapshot of the state), and another is the kernel (which does a context switch so it's handy if the thread has been already saved onto the stack).

I can't think of another case right now, but it must be something this low-level. So I don't think we're getting to a point where there are dozens of users of this API and there will be conflicts in managing ranges. So personally I think "anything goes" here. I wouldn't want to switch back to the mask because I tried that and it was worse readability wise, much harder to understand. I don't have a strong opinion on the other options though, I think anything goes for the few uses we will have of this.

Reply:
I am totally ok 👍🏼